### ISSUE TITLE
Texture swizzling improvements

### ISSUE THREAD

**Issue by billhollings** [KhronosGroup/MoltenVK#306]
@cdavis5e 

Thanks very much for contributing the texture swizzling enhancement. That was a huge endeavour and a significant enhancement. Your contribution was most appreciated! Nice work!

In reviewing the improvements...and making them compatible with caching...I had a couple of design questions I thought I'd raise:

1. The auxiliary `MTLBuffer` is held by the `MVKPipeline`. But that one pipeline might be used in several renderings using different textures. This might set up a race condition where the GPU has not consumed the swizzle data for one texture before the CPU overwrites it with swizzle data for a different texture used in a later draw call. 

     I'm wondering if we could use `MVKCommandEncoder::setVertexBytes()`, etc. to dynamically set the swizzle content instead...similar to how `MVKPushConstantsCommandEncoderState` does. The `SPIRV-Cross` interaction would stay the same.

2. Texture swizzle slots between vertex and fragment shaders seem to overlap in the auxiliary buffer. Does this work if both the vertex and fragment shaders use textures with different swizzles?

3. Currently `swizzle_texture_samples` is always on for any shader that samples a texture. Shall we add a setting to `MVKConfiguration` to control this...to allow apps that know they don't need swizzling to turn it off to save on shader complexity and improve shader performance?

If these suggestions make sense to your overall design for texture swizzling...I can implement them.

**Comment by cdavis5e** [comment]
> 1. The auxiliary `MTLBuffer` is held by the `MVKPipeline`. But that one pipeline might be used in several renderings using different textures. This might set up a race condition where the GPU has not consumed the swizzle data for one texture before the CPU overwrites it with swizzle data for a different texture used in a later draw call.

I see. Since this is state that depends on the bound descriptors (and later, the current draw parameters), it's not really viable to keep it in an `MTLBuffer` owned by the pipeline and shared between multiple draws.

>    I'm wondering if we could use `MVKCommandEncoder::setVertexBytes()`, etc. to dynamically set the swizzle content instead...similar to how `MVKPushConstantsCommandEncoderState` does. The `SPIRV-Cross` interaction would stay the same.

We may be forced to do that. One concern is that the space available with that method is limited: Apple's docs (which are probably accurate for iOS; dunno about Mac) state that this method "is appropriate only for small, one-time-use data buffers. Create a `MTLBuffer` object instead if your buffer exceeds 4 KB in length..." It's not a concern now, since the 512 bytes needed to represent the component mappings for all 128 textures allowed on Mac should fit easily inside that limit, but it will become one when it comes time to support the `VK_EXT_descriptor_indexing` extension, which mandates a 500000-image *minimum* limit. (This whole scheme needs rethinking for that, since we'll need to use indirect argument buffers for that.)

> 2. Texture swizzle slots between vertex and fragment shaders seem to overlap in the auxiliary buffer. Does this work if both the vertex and fragment shaders use textures with different swizzles?

This will almost certainly break in that case. It is now obvious to me that each shader stage that needs an auxiliary buffer needs its *own* auxiliary buffer. (D'oh!)

> 3. Currently `swizzle_texture_samples` is always on for any shader that samples a texture. Shall we add a setting to `MVKConfiguration` to control this...to allow apps that know they don't need swizzling to turn it off to save on shader complexity and improve shader performance?

It doesn't help my employer's case--I doubt programs designed for Windows are going to be checking for a Mac/iOS-specific extension--but I agree this is probably a worthwhile optimization for programs that we know won't need this.

> 
> If these suggestions make sense to your overall design for texture swizzling...I can implement them.

Feel free. Unfortunately, I'm not going to have much time to work on this for the next couple of weeks. My employer has other tasks for me to do. To be honest, I was able to submit the original change because I already had it ready to go. Since it required a SPIRV-Cross update (and therefore, a bump to the MoltenVK version), I wanted to withhold it until I was done updating SPIRV-Cross to support geometry and tessellation shaders. Then I could update SPIRV-Cross once to pull in those changes as well. But, since my free time is limited for now, I probably won't be finishing that for another few weeks.

**Comment by billhollings** [comment]
Sounds good. I'll look into the changes suggested above.

>but it will become one when it comes time to support the VK_EXT_descriptor_indexing extension, which mandates a 500000-image minimum limit

When this becomes necessary...MoltenVK can allocate pools of `MTLBuffers` of various sizes for temporary use. See `MVKCommandEncodingPool::acquireMTLBufferAllocation()` and it's use in `MVKCommandEncoder::copyToTempMTLBufferAllocation()`.

**Comment by billhollings** [comment]
@cdavis5e @danginsburg (and others)

**TL;DR:** Should texture swizzling via shaders be _**on**_ or _**off**_ by default? And should we set this via an environment variable to allow any game to be easily ported with either requirement?

--------------

I'd like to get your opinions.

> It doesn't help my employer's case--I doubt programs designed for Windows are going to be checking for a Mac/iOS-specific extension--but I agree this is probably a worthwhile optimization for programs that we know won't need this.

Some games rely on texture swizzling...and to _**ease porting**_ we should maximize compatibility and turn texture swizzling via shaders _**on**_. Other games ignore texture swizzling...and to _**ease porting**_ we should maximize performance by turning texture swizzling via shaders _**off**_.

I'm looking at several couple of options for flipping it on or off:

1. Dynamically at pipeline compile via a new Vulkan flag in the `VkPipelineShaderStageCreateFlags` bit mask. This would allow a single game to choose when and where to turn it on at the most granular level possible.

2. Statically via a new config boolean in `MVKConfiguration` via `vkGet/SetMoltenVKConfigurationMVK`.

3. Statically via a runtime environment variable...to allow different games to turn it on or off via a launch script without having to special-case a MoltenVK-specific API call.

Any preferences? We could conceptually combine all three options.

Also...should texture swizzling via shaders be turned on or off by default? This seems to be a basic compatability-vs-performance question...but we need to pick one.

I've heard both arguments...compatibility so porting is seamless for those games that do rely on swizzling...and performance...since there are many games that never use texture swizzling...and would like to just run as fast as possible after being ported.

Thoughts and suggestions, please?

**Comment by danginsburg** [comment]
> Thoughts and suggestions, please?

I can see both arguments.  I will say that Artifact paid a 2.5ms penalty on AMD D500 @ 1080p for the texture swizzle behavior which it does not need.  Other games may experience similar cost, which does paint MoltenVK in a negative light for the default behavior.

In a perfect world, in addition to configuration options, MoltenVK could only pay the cost on swizzles once it actually encounters a case where it's needed.  The tough thing here is that it's a cost paid by everyone all the time.  I'm assuming this option is not possible though or you already would have done it that way.

So that being the case, @billhollings feel free to make the call as you see fit.  If you want my vote it would be:
1. Off by default
2. Clear MoltenVK error warning messages when an unsupported swizzle is encountered saying "hey I hit a swizzle I can't support by default, set this configuration option to support it"
3. In the Vulkan Portability Initiative we make this a configuration option so that there is a standardized way to enable it once that spec is done.

**Comment by aerofly** [comment]
I would turn it off by default.

The new texture swizzle enhancement has a performance penalty even if not being used at all. In my eyes it's a waste of CPU / GPU performance that could be used in other places.

Our games are usually pixel shader limited already, so adding this specific feature adds even more overhead.

Since Mac OS devices have on average much slower GPU's, one should try to minimize those unnecessary performance hits. We already have a hard time getting performance right on Mac OS due to the big hardware differences.

So I would opt for 'Statically via a new config boolean in MVKConfiguration...'

**Comment by billhollings** [comment]
@cdavis5e @danginsburg @aerofly 

Thanks for the feedback. Okay...when I make the changes above...I'll turn it off by default...and add text to the swizzle error that mentions it can be turned on.

Is there any merit to the idea of allowing it to be set via an environment variable setting...say from within a script that launches the game on macOS? 

That way...a single, default-compiled `libMoltenVK.dylib` could support different settings for different games...and the game does not have to special-case a platform-specific config API call in the code base.

It seems to me that would minimize porting code changes...and make it easier to change from one option to the other based on runtime environment.

**Comment by cdavis5e** [comment]
I'm in favor of there being an environment variable.

**Comment by danginsburg** [comment]
Envvar is fine by me too.

**Comment by grovesNL** [comment]
I think for Vulkan Portability it would be preferable to avoid an environment variable and use an extension instead -- otherwise all other portability implementations will have to copy this environment variable setup in order to be compatible. This is already a bit unusual because we've implemented `MVK*` extensions.

**Comment by billhollings** [comment]
@grovesNL 

Thanks...yeah...I've raised this with the Vulkan portability TSG as well...so we can standardize.

**Comment by billhollings** [comment]
PR #450 add config settings for texture swizzling...and disables it by default.

**Comment by MarkCallow** [comment]
@billhollings , I'm trying to use swizzles. I set the `components` field of my VkImageViewCreateInfo to the swizzle I want. I then got the warning described here which led me to set the environment variable. The warning disappeared but the swizzle isn't happening. I have an R8_UNORM texture, am setting a swizzle of RRR1 but the texture is displayed as red. Is there something else I need to do to enable swizzling?
